Automated Solvation Shell Analysis

cadeduckworth opened the pull request.

cadeduckworth committed:
developing back end run methods

* separating functionality for use with _single_universe without iterating frames
* retaining functionality for use with _single_frame

cadeduckworth committed:
condensed backend run methods

* try/except pattern condensed to one run method
* single run method implemented for _single_universe and _single_frame
* functionality of original run method retained, while removing frame iteration for _single_universe usage
* progress bar still active when using _single_universe, but no frame iteration occurs

cadeduckworth committed:
Initial adding of automation functionality to dihedral.py
directory_paths: finds MDPOW project data and returns pandas DataFrame
directory_iteration: takes pandas DataFrame or csv and iterates automated_dihedral_analysis over MDPOW project directories
automated_dihedral_analysis: group of functions that includes automation scripts for DihedralAnalysis, automatic dihedral atom group selection, seaborn violin plots function, and other required functions
Future changes will eliminate redundancies following initial testing

cadeduckworth committed:
moved directory_paths function to reside within the DihedralAnalysis class to test possible ways to call automation-related functions

cadeduckworth committed:
Addressed change requests by Oliver
removed raise
added docstrings to describe use of NotImplementedError
responded with reason for removing _conclude_universe()
next step: updating html documentation following successful testing

cadeduckworth committed:
fixed indentation error for _single_frame() block

cadeduckworth committed:
reverting previous tab issue, necessary for normal function

cadeduckworth committed:
fixed errors previously reverted for testing
fixed expected indent block error
removed raise
removed _conclude_universe()

cadeduckworth committed:
Merge branch 'develop' into ensemble_run_update

cadeduckworth committed:
restored original dihedral.py file

cadeduckworth committed:
created automation/dihedral subdirectory under mdpow/analysis/ to house automation for DihedralAnalysis
automated_dihedral_analysis.py
directory_iteration.py
directory_paths.py

cadeduckworth committed:
updated docstrings and added examples for directory_paths.py

cadeduckworth committed:
added docstrings and examples for directory_iteration.py

cadeduckworth committed:
added initial draft of docstrings and examples for all functions contained within automated_dihedral_analysis.py

cadeduckworth committed:
added the option to save the DihedralAnalysis results DataFrame as csv files in provided parent directory

cadeduckworth committed:
removed redundant DataFrame saving pattern from automated_dihedral_analysis

cadeduckworth committed:
Merge remote-tracking branch 'origin/ensemble_run_update' into automated-dihedral-analysis
updating with changes for backend and testing from PR #216 ensemble_run_update

cadeduckworth committed:
merged updates from PR #216 and PR #218

orbeckst committed:
Merge branch 'develop' into automated-dihedral-analysis

orbeckst committed:
Merge branch 'develop' into automated-dihedral-analysis

cadeduckworth committed:
resolved conflicts to merge updates in from develop from PR#216

cadeduckworth committed:
adding data for testing automated dihedral analysis

cadeduckworth committed:
added starting framework for testing automated dihedral analysis

cadeduckworth committed:
reorganized automation directory into workflows directory

cadeduckworth committed:
added preliminary simple core tests for automated dihedral analysis

cadeduckworth committed:
initial reformatting of existing docs for sphinx markup compatibility
*incomplete, pushing to test
*source and html doc additions
*changed all instances of datadir to dirname for consistency with DihedralAnalysis
*all changes pertain to new workflows modules

cadeduckworth committed:
fix errors in docs for workflows

cadeduckworth committed:
added imports for automated dihedral analysis tests

cadeduckworth committed:
added init file for workflows module

cadeduckworth committed:
added logging functionality for workflows modules

cadeduckworth committed:
sphinx markup corrections for workflow modules

cadeduckworth committed:
redo of reduced testing data for GAFF SM25 from original full dataset, reduced original by factor of 100 this time, timestep is now 50ps, amount of frames and total length of simulation preserved

cadeduckworth committed:
sphinx docs automodule test

cadeduckworth committed:
sphinx docs test base.txt

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
edit docs

cadeduckworth committed:
edit docs

cadeduckworth committed:
edit docs

cadeduckworth committed:
edit docs

cadeduckworth committed:
edit docs

cadeduckworth committed:
update py.path to pathlib in test_automated_dihedral_analysis

cadeduckworth committed:
update reference testing values for new reduced dataset, test_automated_dihedral_analysis

cadeduckworth committed:
change syntax in test_automated_dihedral_analysis to compare objects instead of strings

cadeduckworth committed:
update assert_almost_equal to pytest.approx

cadeduckworth committed:
bounce redundant function calls to a pytest.fixture for use with multiple tests

cadeduckworth committed:
add bz2 compression for dataframe storage of automated dihedral analysis data

cadeduckworth committed:
Update ci.yaml

cadeduckworth committed:
added functionality for user input of alternative SMARTS string selection for automated dihedral analysis, including docs and tests

cadeduckworth committed:
Merge branch 'automated-dihedral-analysis' of github.com:Becksteinlab/MDPOW into automated-dihedral-analysis

cadeduckworth committed:
add tests for automated directory iteration of dihedral analysis, workflows/base.py

cadeduckworth committed:
add tests for saving dataframes and saving figures, which covers the full automation of automated dihedral analysis

cadeduckworth committed:
add test for directory iteration of fully automated dihedral analysis
* cleanup of aux files is needed
* updated .gitignore file is neded
* added SM26 reduced dataset for iteration test

cadeduckworth committed:
moved location of rdkit install used for testing/pytest

cadeduckworth committed:
remove misplaced hbonds documentation

cadeduckworth committed:
relocate test data for workflows and change test scripts accordingly

cadeduckworth committed:
removed unnecessary .ipynb, .lock, and .npz files

cadeduckworth committed:
simplified keyword specification for default and user input of SMARTS string selection

cadeduckworth committed:
fixed plotting issue for solvents (name and order)

cadeduckworth committed:
variable name issue in testing module, user_SMARTS -> SMARTS

cadeduckworth committed:
Delete dir.csv

cadeduckworth committed:
import syntax

cadeduckworth committed:
sphinx docs

cadeduckworth committed:
Merge branch 'automated-dihedral-analysis' of github.com:Becksteinlab/MDPOW into automated-dihedral-analysis
change made directly on github.com caused diversion

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
moving workflows location, subsequent doc changes

cadeduckworth committed:
import changed to reflect moving workflows module

cadeduckworth committed:
docs

cadeduckworth committed:
add requirements for sphinx build

cadeduckworth committed:
add requirements for sphinx build

cadeduckworth committed:
add requirements for sphinx build

cadeduckworth committed:
add requirements for sphinx build

cadeduckworth committed:
add requirements for sphinx build

cadeduckworth committed:
docs

cadeduckworth committed:
docs

cadeduckworth committed:
docs structure

cadeduckworth committed:
consistent keyword names in tests

cadeduckworth committed:
imports and docs

cadeduckworth committed:
imports

cadeduckworth committed:
imports

cadeduckworth committed:
generalizing automation base module for use with other analyses

cadeduckworth committed:
initialize PR for automated-solvation-shell module

cadeduckworth committed:
initial script for automated solvation shell analysis

cadeduckworth committed:
updated automated solvation module, incomplete, issue with compression

cadeduckworth committed:
fixed automated solvation shell analysis module
ASA works as is
workflows base module needs adjustment for proper use

cadeduckworth committed:
docs and func names

cadeduckworth committed:
cleanup and remove workflows base module/move to new PR

orbeckst commented:
Lots of conflicts to resolve, @cadeduckworth !

cadeduckworth committed:
resolving conflicts

cadeduckworth committed:
add new version of testing resources

cadeduckworth committed:
correct module name

cadeduckworth committed:
start test module for solvations

cadeduckworth committed:
change data obtained by SolvationAnalysis

cadeduckworth commented:
@orbeckst 
The following is a brief overview of the new manual `EnsembleAnalysis` workflow based on the changes I made to the original `SolvationAnalysis`: (the changes to `mdpow.analysis.solvation` are in the most recent commit`)

I made changes that allow a user to input a max distance, and obtain all of the solute-solvent interactions within that distance (distance and atom/atomgroup info included), per molecule, per solvent, per interaction, per lambda.

This is a very rough draft, especially for the plot, but the workflow, results DataFrame, and FacetGrid get the main idea across for what I am going for.

SAMPL9 Data, FF = GAFF, Molecule = S08

Workflow:
![Screen Shot 2023-04-05 at 12 37 38 PM](https://user-images.githubusercontent.com/104414304/230161864-b262998d-376e-41e3-a465-fc6679338805.png)

Results DF: (it might be better to break the `*_ix` lists up into more columns)
![Screen Shot 2023-04-05 at 12 37 55 PM](https://user-images.githubusercontent.com/104414304/230161957-cff2cadf-f627-4041-b48b-972349687a8f.png)

Figure (FacetGrid): (please excuse the quality, I am trying to work out the kinks in the `.map()` function)
![Screen Shot 2023-04-05 at 12 30 41 PM](https://user-images.githubusercontent.com/104414304/230162075-c6165bdf-83f6-466b-9cd2-ffe7565660c0.png)

orbeckst commented:
The manual workflow as shown is a good mini tutorial and looks already very good. 

Presenting the data will be more of a challenge—the distance KDEs are naturally cut off at the cutoff, which makes it look a bit weird/incomplete. And once the solute disappears, the N becomes quite big and swamps the small numbers. Perhaps try a log plot?? Or normalize??? Also, are the large numbers compatible with what you get for a sphere of radius with the cutoff at the density of water or octanol/other solvent?

orbeckst commented:
The distance histograms are interesting. You can see how molecules start to overlap with the disappearing solute.

cadeduckworth committed:
name top-level automated solvation function and complete addition to workflows registry/registry docs

cadeduckworth commented:
@orbeckst 
Although there is a *large* step size specified, from a theoretical perspective I interpreted the violins as a stepwise visual contribution to confirmation that things are going according to plan in the context of turning off intermolecular interactions. I might be thinking about it wrong, but the solute or solvent is essentially being electrostatically and otherwise alchemically removed from the system which is positively indicated by the increased equilibration of solvent molecules, right? On that note, I think seeing that there is not much change when turning of Coulombic interactions for neutral solute/solvent is a good indication, and then seeing the gradual increase of solvent molecules at a decreased distance when turning off VDW in that sense is also a good indication of sufficient sampling, *IFF*, this holds up for other cases with more frames included.

Perhaps I should run a simulation with one of the charged molecules that was excluded for logP calculations from one of the SAMPL sets and see what happens if I can get GROMACS to comply?

cadeduckworth commented:
@orbeckst

I will look into approximately homogenous equilibration of solvents to try to confirm the data for more completely sampled datasets when I get the chance (those lambdas for VDW closer to 1, when the solute is essentially *removed*). I apologize if I did not use the correct terms there, it has been a while since I covered equilibrium chemistry, but I think I understand what you are getting at.

cadeduckworth commented:
@orbeckst 

I ran a more complete analysis of the same SAMPL9 GAFF molecule (S08) with a step size of 500 and cleaned up the plots a bit.
This is the product of what I have been playing with using `cairosvg` and `svgutils` instead of `sns.FacetGrid`, which, outside of `.catplots`, when initialized directly, pretty much requires that axes be shared for the same data being analyzed, which is exactly the same as sharing axes in `.catplots`. 

So, this is the SVG converted to pdf. The width ratios are slightly off, which affects the placement of the Mol object, but I cleaned up the scale on the y-axes as well as the height/width aspect ratio. Fixing the axes/scales and including much more data seems to give a better look at what is going on.

If we do eventually go with a stacked plot scheme, then I will reserve less of the right side and reduce it to a single Mol object and legend to make it look better.


[S08.pdf](https://github.com/Becksteinlab/MDPOW/files/11166519/S08.pdf)

cadeduckworth commented:
From earlier this morning before making changes to `SolvationAnalysis`:

Using the original `SolvationAnalysis` class, which summed the number of solvent molecules within a specified first and second cutoff distance, separately, I made a time series of the number of solvent molecules.
The concept does not make much sense since each lambda simulation starts from the same relaxed equilibrated MD sim(?), but if I can map the time series to the lambda values then it might be useful as an option later on for something. Note: this is using very under sampled data, large step size, but also shows 95% CI

![Screen Shot 2023-04-06 at 12 00 55 AM](https://user-images.githubusercontent.com/104414304/230276200-e835123a-85a4-4344-ab3b-6d4ed75b4f31.png)

orbeckst commented:
For the S08 data, the number of solvent atoms is initially 60,000 — there aren't that many waters in the box, are they? 

Can you do the following calculation: Given the standard density of water at the simulation conditions, how many water molecules would you expect to find in a volume roughly equivalent to (1) the solute, (2) the solute with its surface blown up by 4 Å? A rough physicist estimate will suffice as a start, e.g., approximating the molecule as a cylinder or sphere...

Then compare the estimate to the numbers from solvation analysis.

orbeckst commented on mdpow/analysis/solvation.py:
```diff
@@ -56,21 +56,38 @@ def __init__(self, solute: EnsembleAtomGroup, solvent: EnsembleAtomGroup, distan
 
     def _prepare_ensemble(self):
-        self._col = ['distance', 'solvent', 'interaction',
-                     'lambda', 'time', 'N_solvent']
```
Are you now returning distances? If so, better write a new class and leave the old one as-is.

cadeduckworth replied:
@orbeckst 
I agree, because they are doing different things. Can they both exist as separate classes within the same module?

orbeckst commented:
I am not sure if the time series contain more interesting information than the averages or distributions — at least in equilibrium. Time series in equilibrium are good as a diagnostic to check if the system is likely sampling equilibrium: if the fluctuations are around a stable mean then it might be equilibrium...

orbeckst commented:
I don't think that we need to add the molecular structure next to the solvation plots unless you want to highlight specific atoms. Otherwise I'd expect a molecular structure to be included in an overview panel and the reader can use the ID to cross-reference.

cadeduckworth commented:
@orbeckst 
I am working on some calculations/estimates, but for clarification, the 'counts' bar plot is counting each solvent atom that is within the cutoff distance near any solute atom non-exclusively, so multiple solvent atom - solute atoms distances are being recorded and counted, meaning that solvent atom A near solute atom B *and* solute atom C are counted separately. I believe this is why the numbers are so high. There could be an easy way to count each solvent atom only once with a minor adjustment to the script. Additionally, we could still record all of the data, and only count each solvent molecule once in the plot by only counting the unique `resid`s.

orbeckst commented:
Each atom should only be counted once. There's a single closest interaction between a solvent atom and the solute (barring coincidental equidistant triangles).

orbeckst commented:
@cadeduckworth is this something you'd want to complete for a MDPOW paper?

Can you resolve the conflicts so that one can look at the changes in the context of the latest version of the code?
